### PR TITLE
EnPp (Hiploop) OK and documented, object_pp documented

### PR DESCRIPTION
**Pull request by tom-overton**
This actor definitely feels like something that was fairly rushed, and as such I've grown a bit sick of it. I tried as best as I could to document stuff, but if something is confusing, I'm happy to add more documentation. I just figured it would be best to put this out there to more easily find the confusing parts.

### CODE REVIEW

File: src/overlays/actors/ovl_En_Pp/z_en_pp.c [hunk 1]
```diff
@@ -64,116 +127,1500 @@ const ActorInit En_Pp_InitVars = {
...
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Pp/func_80B20668.s")
+    if (this->floorPolyForCircleShadow != NULL) {
+        if ((this->action != EN_PP_ACTION_MASK_DETACH) && (this->action < EN_PP_ACTION_DEAD)) {
+            OPEN_DISPS(play->state.gfxCtx);
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Pp/EnPp_Update.s")
+            func_8012C448(play->state.gfxCtx);
+            gDPSetPrimColor(POLY_XLU_DISP++, 0, 0, 0x00, 0x00, 0x00, 0xFF);
```
Comment (jpburnett) [inline src/overlays/actors/ovl_En_Pp/z_en_pp.c:1610]:
```suggestion
            gDPSetPrimColor(POLY_XLU_DISP++, 0, 0, 0, 0, 0 255);
```
I believe these can be decimal

**Review by jpburnett — COMMENTED** [review]
Awesome work!

File: src/overlays/actors/ovl_En_Pp/z_en_pp.c [hunk 2]
```diff
@@ -64,116 +127,1500 @@ const ActorInit En_Pp_InitVars = {
...
+
+    if (EN_PP_GET_ATTACK_RANGE(&this->actor) == 0) {
+        this->attackRange = 200.0f;
+    } else {
+        this->attackRange = EN_PP_GET_ATTACK_RANGE(&this->actor) * 40.0f;
+    }
+
+    this->chargesInStraightLines = EN_PP_DOES_CHARGE_IN_STRAIGHT_LINES(&this->actor);
+    if (this->chargesInStraightLines) {
+        this->actor.params &= 1;
```
[the remark this answers is not in the document]
Comment (tom-overton) [reply src/overlays/actors/ovl_En_Pp/z_en_pp.c:252]: Hmm, I'm not so sure about this one. I'm not sure that using enum values as bitmasks like this makes total sense. Let's say that you have a params of 0x0100. After doing `this->actor.params &= 1`, your params will be equal to `EN_PP_TYPE_MASKED`; that seems like a rather unintuitive result given that `NO_MASK` is written there. What do you think?

[the remark this answers is not in the document]
Comment (jpburnett) [reply src/overlays/actors/ovl_En_Pp/z_en_pp.c:252]: Uh...this was a test and you passed! (just kidding, I didn't see the `&`, I was going through looking at the usage of params and saw this case), I think leaving it as is with the mask is fine.

**Review by jpburnett — APPROVED** [review]

**Review by AngheloAlf — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**